theano/scalar/basic.py

```diff
@@ -2399,9 +2399,19 @@ def c_code(self, node, name, inputs, outputs, sub):
 
 
 class Neg(UnaryScalarOp):
+    # We can use numpy.negative here, because even if it gives unexpected
+    # results on Boolean arrays, it will be passed other dtypes as Theano
+    # does not have a Boolean type for tensors.
     nfunc_spec = ('negative', 1, 1)
 
     def impl(self, x):
+        # We have to make sure x is not a numpy.bool_, because
+        # `-numpy.bool_(True)` is `False` (we want 0), and
+        # `-numpy.bool_(False)` is `True` (we want 1).
+        # This happens for Composite, as the intermediate results are not
+        # casted in the dtype of the intermediate variable in general.
+        if isinstance(x, numpy.bool_):
+            x = numpy.int8(x)
```

Comment:
There is a test that was failing in DebugMode, but we could add a new one.
As I say in the comment, when we have a composite Op, the intermediate values are not casted into the dtype of the intermediate variable. In that case, we had something like `cast{float32}(-(i0 > i1))`, so the intermediate result of `i0 > i1` was a single `numpy.bool_` scalar, and then `Neg.impl` was called on it.